File: backend/src/main/java/com/tailorstudio/app/web/AuthOtpController.java
package com.tailorstudio.app.web;

import com.tailorstudio.app.dto.LoginOtpResendRequest;
import com.tailorstudio.app.dto.LoginPasswordChallengeRequest;
import com.tailorstudio.app.dto.OtpEmailRequest;
import com.tailorstudio.app.dto.OtpLoginVerifyRequest;
import com.tailorstudio.app.dto.OtpVerifyRequest;
import com.tailorstudio.app.dto.PasswordResetRequest;
import com.tailorstudio.app.security.StudioUserDetails;
import com.tailorstudio.app.service.MePayloadService;
import com.tailorstudio.app.service.OtpAuthService;
import jakarta.validation.Valid;
import org.springframework.http.HttpStatus;
import org.springframework.http.ResponseEntity;
import org.springframework.web.bind.annotation.PostMapping;
import org.springframework.web.bind.annotation.RequestBody;
import org.springframework.web.bind.annotation.RequestMapping;
import org.springframework.web.bind.annotation.RestController;

import java.time.Instant;
import java.time.temporal.ChronoUnit;
import java.util.Map;

@RestController
@RequestMapping("/api/auth/otp")
public class AuthOtpController {

    private final OtpAuthService otpAuthService;
    private final MePayloadService mePayloadService;

    public AuthOtpController(OtpAuthService otpAuthService, MePayloadService mePayloadService) {
        this.otpAuthService = otpAuthService;
        this.mePayloadService = mePayloadService;
    }

    /** Millisecond precision only — avoids JS Date parse issues with nanosecond ISO-8601 from {@link Instant#toString()}. */
    private static String expiresAtIso(Instant instant) {
        return instant.truncatedTo(ChronoUnit.MILLIS).toString();
    }

    @PostMapping("/login/challenge")
    public ResponseEntity<Map<String, Object>> loginChallenge(@Valid @RequestBody LoginPasswordChallengeRequest body) {
        try {
            var r = otpAuthService.startLoginWithPassword(body.email(), body.password());
            return ResponseEntity.ok(
                    Map.of(
                            "ok",
                            true,
                            "expiresAt",
                            expiresAtIso(r.expiresAt()),
                            "pendingToken",
                            r.pendingTokenPlain(),
                            "staticOtp",
                            otpAuthService.isStaticOtpBypassEnabled()));
        } catch (OtpAuthService.NoSuchUserException e) {
            return ResponseEntity.status(HttpStatus.NOT_FOUND).body(Map.of("ok", false, "error", "no_account"));
        } catch (OtpAuthService.BadPasswordException e) {
            return ResponseEntity.status(HttpStatus.UNAUTHORIZED).body(Map.of("ok", false, "error", "invalid_credentials"));
        } catch (IllegalStateException e) {
            return ResponseEntity.status(HttpStatus.SERVICE_UNAVAILABLE)
                    .body(Map.of("ok", false, "error", "mail_not_configured", "message", e.getMessage()));
        }
    }

    @PostMapping("/login/resend")
    public ResponseEntity<Map<String, Object>> loginResend(@Valid @RequestBody LoginOtpResendRequest body) {
        try {
            var exp = otpAuthService.resendLoginOtp(body.pendingToken());
            return ResponseEntity.ok(Map.of("ok", true, "expiresAt", expiresAtIso(exp)));
        } catch (OtpAuthService.OtpInvalidException e) {
            return ResponseEntity.status(HttpStatus.BAD_REQUEST).body(Map.of("ok", false, "error", "invalid_pending"));
        } catch (IllegalStateException e) {
            return ResponseEntity.status(HttpStatus.SERVICE_UNAVAILABLE)
                    .body(Map.of("ok", false, "error", "mail_not_configured", "message", e.getMessage()));
        }
    }

    @PostMapping("/login/verify")
    public ResponseEntity<?> loginVerify(@Valid @RequestBody OtpLoginVerifyRequest body) {
        try {
            var out = otpAuthService.verifyLoginOtp(body.email(), body.code(), body.pendingToken());
            var me = mePayloadService.buildFor(new StudioUserDetails(out.user()));
            return ResponseEntity.ok(Map.of("ok", true, "me", me, "accessToken", out.accessToken()));
        } catch (OtpAuthService.OtpInvalidException e) {
            return ResponseEntity.status(HttpStatus.BAD_REQUEST).body(Map.of("ok", false, "error", "invalid_otp"));
        }
    }

    @PostMapping("/forgot/send")
    public ResponseEntity<Map<String, Object>> sendForgot(@Valid @RequestBody OtpEmailRequest body) {
        try {
            var exp = otpAuthService.sendForgotPasswordOtp(body.email());
            return ResponseEntity.ok(Map.of("ok", true, "expiresAt", expiresAtIso(exp)));
        } catch (OtpAuthService.NoSuchUserException e) {
            return ResponseEntity.status(HttpStatus.NOT_FOUND).body(Map.of("ok", false, "error", "no_account"));
        } catch (IllegalStateException e) {
            return ResponseEntity.status(HttpStatus.SERVICE_UNAVAILABLE)
                    .body(Map.of("ok", false, "error", "mail_not_configured", "message", e.getMessage()));
        }
    }

    @PostMapping("/forgot/verify")
    public ResponseEntity<Map<String, Object>> verifyForgot(@Valid @RequestBody OtpVerifyRequest body) {
        try {
            String token = otpAuthService.verifyForgotOtpAndIssueResetToken(body.email(), body.code());
            return ResponseEntity.ok(Map.of("ok", true, "resetToken", token));
        } catch (OtpAuthService.OtpInvalidException e) {
            return ResponseEntity.status(HttpStatus.BAD_REQUEST).body(Map.of("ok", false, "error", "invalid_otp"));
        }
    }

    @PostMapping("/forgot/reset")
    public ResponseEntity<Map<String, Object>> resetPassword(@Valid @RequestBody PasswordResetRequest body) {
        try {
            otpAuthService.resetPasswordWithToken(body.resetToken(), body.newPassword());
            return ResponseEntity.ok(Map.of("ok", true));
        } catch (OtpAuthService.OtpInvalidException e) {
            return ResponseEntity.status(HttpStatus.BAD_REQUEST).body(Map.of("ok", false, "error", "invalid_reset_token"));
        } catch (IllegalArgumentException e) {
            return ResponseEntity.badRequest().body(Map.of("ok", false, "error", "validation", "message", e.getMessage()));
        }
    }
}

File: server/src/controllers/authOtpController.ts
import type { Request, Response } from 'express'
import * as otp from '../services/otpAuthService.js'
import { buildMe } from '../services/mePayloadService.js'
import { expiresAtIso } from '../utils/apiJson.js'
import type { UserDoc } from '../models/User.js'

export async function loginChallenge(req: Request, res: Response) {
  try {
    const r = await otp.startLoginWithPassword(req.body.email, req.body.password)
    res.json({
      ok: true,
      expiresAt: expiresAtIso(r.expiresAt),
      email: r.email,
      staticOtp: otp.isStaticOtpBypassEnabled(),
    })
  } catch (e) {
    const m = e instanceof Error ? e.message : ''
    if (m === 'no_account') return res.status(404).json({ ok: false, error: 'no_account' })
    if (m === 'invalid_credentials') return res.status(401).json({ ok: false, error: 'invalid_credentials' })
    if (m.includes('not configured')) {
      return res.status(503).json({ ok: false, error: 'mail_not_configured', message: m })
    }
    res.status(400).json({ ok: false, error: m })
  }
}

export async function loginResend(req: Request, res: Response) {
  try {
    const exp =
      req.body.password != null && String(req.body.password).length > 0
        ? await otp.resendLoginOtpWithPassword(req.body.email, req.body.password)
        : await otp.resendLoginOtp(req.body.pendingToken)
    res.json({ ok: true, expiresAt: expiresAtIso(exp) })
  } catch (e) {
    const m = e instanceof Error ? e.message : ''
    if (m === 'no_account') return res.status(404).json({ ok: false, error: 'no_account' })
    if (m === 'invalid_credentials') return res.status(401).json({ ok: false, error: 'invalid_credentials' })
    if (m === 'invalid_pending') return res.status(400).json({ ok: false, error: 'invalid_pending' })
    if (m.includes('not configured')) {
      return res.status(503).json({ ok: false, error: 'mail_not_configured', message: m })
    }
    res.status(400).json({ ok: false, error: m })
  }
}

async function respondLoginSuccess(res: Response, out: Awaited<ReturnType<typeof otp.completeLoginWithPasswordAndOtp>>) {
  const me = await buildMe(out.user as UserDoc)
  res.json({ ok: true, me, accessToken: out.accessToken })
}

function loginVerifyError(res: Response, e: unknown) {
  const m = e instanceof Error ? e.message : ''
  if (m === 'no_account') return res.status(404).json({ ok: false, error: 'no_account' })
  if (m === 'invalid_credentials') return res.status(401).json({ ok: false, error: 'invalid_credentials' })
  if (m === 'invalid_pending') return res.status(400).json({ ok: false, error: 'invalid_pending' })
  if (m === 'user_not_found') return res.status(400).json({ ok: false, error: 'user_not_found' })
  if (m === 'Business not found') {
    return res.status(503).json({ ok: false, error: 'business_not_found', message: m })
  }
  if (m === 'Document missing numeric id') {
    return res.status(503).json({ ok: false, error: 'legacy_data', message: m })
  }
  const dev = process.env.NODE_ENV !== 'production'
  return res.status(400).json({
    ok: false,
    error: m === 'invalid_otp' ? 'invalid_otp' : m || 'invalid_otp',
    ...(dev && m ? { message: m } : {}),
  })
}

export async function loginComplete(req: Request, res: Response) {
  try {
    const out = await otp.completeLoginWithPasswordAndOtp(
      req.body.email,
      req.body.password,
      req.body.code,
    )
    await respondLoginSuccess(res, out)
  } catch (e) {
    loginVerifyError(res, e)
  }
}

export async function loginVerify(req: Request, res: Response) {
  try {
    const out =
      req.body.password != null && String(req.body.password).length > 0
        ? await otp.completeLoginWithPasswordAndOtp(req.body.email, req.body.password, req.body.code)
        : await otp.verifyLoginOtp(req.body.email, req.body.code, req.body.pendingToken)
    await respondLoginSuccess(res, out)
  } catch (e) {
    loginVerifyError(res, e)
  }
}

export async function forgotSend(req: Request, res: Response) {
  try {
    const exp = await otp.sendForgotPasswordOtp(req.body.email)
    res.json({ ok: true, expiresAt: expiresAtIso(exp) })
  } catch (e) {
    const m = e instanceof Error ? e.message : ''
    if (m === 'no_account') return res.status(404).json({ ok: false, error: 'no_account' })
    if (m.includes('not configured')) {
      return res.status(503).json({ ok: false, error: 'mail_not_configured', message: m })
    }
    res.status(400).json({ ok: false, error: m })
  }
}

export async function forgotVerify(req: Request, res: Response) {
  try {
    const resetToken = await otp.verifyForgotOtp(req.body.email, req.body.code)
    res.json({ ok: true, resetToken })
  } catch {
    res.status(400).json({ ok: false, error: 'invalid_otp' })
  }
}

export async function forgotReset(req: Request, res: Response) {
  try {
    await otp.resetPasswordWithToken(req.body.resetToken, req.body.newPassword)
    res.json({ ok: true })
  } catch (e) {
    const m = e instanceof Error ? e.message : ''
    if (m === 'invalid_reset_token') return res.status(400).json({ ok: false, error: 'invalid_reset_token' })
    if (m.includes('8 characters')) {
      return res.status(400).json({ ok: false, error: 'validation', message: m })
    }
    res.status(400).json({ ok: false, error: m })
  }
}
